### PR TITLE
(std/log): Async buffered file logging

### PR DESCRIPTION
**Pull request by cknight**
This PR introduces two improvements to the performance of the logging FileHandler (and RotatingFileHandler).  The first is the use of an async queue to handle logs (making logging for these handlers an async process) and the second is the use of a buffered writer to improve log file write throughput.  The bulk of logic comes in the form of a new `Queue` class which I ended up refactoring to be a top level citizen in a new `std/collections` folder as I believe it to be of use to others.

**Performance improvements**

Significant performance improvements were observed under two different scenarios tested.  The first was running a web server under load (using autocannon) which, for each request, logged 2 log statements to a FileHandler.  A modest boost of 19% in requests processed per second was observed with this change.

The second scenario tested was a for loop which logged once per iteration to a file.  Performance gains of around 500% were observed (in loop iterations per second).

Additional arbitrary tests were also carried out which simulated a slow write (a remote disk for example).  Introducing a 2ms delay on each write had a massive negative impact on the existing logging solution (each log statement being a blocking single sync write) showing this PR solution processing 4,500% more requests per second in the web server.  Even bigger massive gains were also observed in the for loop logging.

There are no logging API changes, and the code handles flushing of the buffer automatically.

cc @bartlomieju

### CODE REVIEW

File: std/log/handlers_test.ts [hunk 1]
```diff
@@ -22,6 +22,18 @@ class TestHandler extends BaseHandler {
...
 }
 
+/** Messages are put into a queue and processed in a for await async loop.  This
+ * function will pause test execution and let the next event loop run to process
+ * the remaining items on the queue, and then flush the buffer to allow the test
+ * to check the output in the destination file.
+ */
+async function nextEventLoopAndFlushLogs(handler: FileHandler): Promise<void> {
+  await new Promise((res) => {
+    setTimeout(res, 0);
```
Comment (ry) [inline std/log/handlers_test.ts:32]: Use `queueMicrotask()`

File: std/log/handlers.ts [hunk 2]
```diff
@@ -99,11 +102,14 @@ interface FileHandlerOptions extends HandlerOptions {
...
 
 export class FileHandler extends WriterHandler {
+  protected _buff!: BufWriterSync;
   protected _file!: File;
   protected _filename: string;
   protected _mode: LogMode;
   protected _openOptions: OpenOptions;
   #encoder = new TextEncoder();
+  #queue: Queue<string> = new Queue<string>();
+  #drainComplete!: Deferred<void>;
```
Comment (ry) [inline std/log/handlers.ts:112]: I think logging output should be done synchronously. Is this patch making that not so?

Comment (cknight) [reply std/log/handlers.ts:112]: @ry Ultimately, all logs are being  output to file synchronously (through `BufWriterSync.writeSync()`).  The queue here is used as a buffer for log messages to be processed in the next event loop.  I'm actually going to close this PR.  I've just had an epiphany on how to radically simplify this solution and achieve the same result.

**Comment by cknight** [comment]
I've come up with a better solution.  Will raise a separate PR for it.

**Pull request closed without merging**